validator/delphi_validator/validate.py

```diff
@@ -0,0 +1,742 @@
+
+            # Check data from a group of dates against recent (previous 7 days, by default)
+            # data from the API.
+            for checking_date in date_list:
+                recent_cutoff_date = checking_date - recent_lookbehind
+                recent_df = geo_sig_df.query(
+                    'time_value <= @checking_date & time_value >= @recent_cutoff_date')
+
+                if recent_df.empty:
+                    self.raised_errors.append(ValidationError(
+                        ("check_missing_geo_sig_date_combo",
+                         checking_date, geo, sig),
+                        None,
+                        "Test data for a given checking date-geo-sig combination is missing"))
+                    continue
+
+                # Reference dataframe runs backwards from the checking_date
+                reference_start_date = checking_date - \
+                    min(semirecent_lookbehind, self.max_check_lookbehind)
+                reference_end_date = recent_cutoff_date - timedelta(days=1)
+                reference_api_df = fetch_api_reference(
+                    self.data_source, reference_start_date, reference_end_date, geo, sig)
+
+                self.check_max_date_vs_reference(
+                    recent_df, reference_api_df, checking_date, geo, sig)
+
+                if self.sanity_check_rows_per_day:
+                    self.check_rapid_change_num_rows(
+                        recent_df, reference_api_df, checking_date, geo, sig)
+
+                if self.sanity_check_value_diffs:
+                    self.check_avg_val_diffs(
+                        recent_df, reference_api_df, checking_date, geo, sig)
+
+            # Keeps script from checking all files in a test run.
+            if self.test_mode:
+                kroc += 1
+                if kroc == 2:
+                    break
+
+        self.exit()
+
+    def exit(self):
+        """
+        If any not-suppressed exceptions were raised, print and exit with non-zero status.
+        """
+        if self.raised_errors:
+            suppressed_counter = 0
+            subset_raised_errors = []
+
+            for val_error in self.raised_errors:
+                # Convert any dates in check_data_id to strings for the purpose of comparing
+                # to manually suppressed errors.
+                raised_check_id = tuple([
+                    item.strftime("%Y-%m-%d") if isinstance(item, (date, datetime))
+                    else item for item in val_error.check_data_id])
+
+                if raised_check_id not in self.suppressed_errors:
+                    subset_raised_errors.append(val_error)
+                else:
+                    self.suppressed_errors.remove(raised_check_id)
+                    suppressed_counter += 1
+
+            print(len(subset_raised_errors), "messages")
+            print(suppressed_counter, "suppressed messages")
+
+            if len(subset_raised_errors) == 0:
```

Comment:
It looks like you can remove this block since it will be handled by the last one anyway